File: app/src/test/java/com/leaguechampions/presenter/SettingsPresenterTest.java
package com.leaguechampions.presenter;

import android.content.SharedPreferences;
import android.view.MenuItem;

import com.leaguechampions.BuildConfig;
import com.leaguechampions.utils.PrefUtils;
import com.leaguechampions.utils.ReflectionUtils;

import org.junit.After;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.Mock;
import org.powermock.api.mockito.PowerMockito;
import org.powermock.core.classloader.annotations.PrepareForTest;
import org.powermock.modules.junit4.PowerMockRunner;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.verifyNoMoreInteractions;
import static org.mockito.Mockito.when;

@RunWith(PowerMockRunner.class)
@PrepareForTest({BuildConfig.class, PrefUtils.class})
public class SettingsPresenterTest {

    private SettingsPresenter presenter;

    private final String fieldBuildType = "buildType";

    @Mock SettingsPresenter.SettingsViewable viewable;
    @Mock SharedPreferences sharedPreferences;

    @Before
    public void setUp() {
        presenter = new SettingsPresenter(sharedPreferences);
        presenter.setViewable(viewable);
    }

    @After
    public void tearDown() {

    }

    @Test
    public void testOnActivityCreated() {
        PowerMockito.mockStatic(PrefUtils.class);
        when(PrefUtils.isMockMode(sharedPreferences)).thenReturn(true);
        ReflectionUtils.setField(presenter, fieldBuildType, "debug");

        presenter.onActivityCreated(null, null);

        verify(viewable).setVersion(BuildConfig.VERSION_NAME);
        verify(viewable).showMockMode(true);
        verifyNoMoreInteractions(viewable);
    }

    @Test
    public void testOnActivityCreated_WhenBuildTypeRelease_OnlySetsVersion() {
        PowerMockito.mockStatic(PrefUtils.class);
        ReflectionUtils.setField(presenter, fieldBuildType, "release");

        presenter.onActivityCreated(null, null);

        verify(viewable).setVersion(BuildConfig.VERSION_NAME);
        verifyNoMoreInteractions(viewable);
    }

    @Test
    public void testOnOptionsItemSelected() {
        MenuItem item = mock(MenuItem.class);
        when(item.getItemId()).thenReturn(android.R.id.home);

        boolean result = presenter.onOptionsItemSelected(item);

        assertThat(result).isEqualTo(true);
        verify(viewable).doFinish();
        verifyNoMoreInteractions(viewable);
    }

    @Test
    public void testOnOptionsItemSelected_WhenItemNotHome_DoNothing() {
        MenuItem item = mock(MenuItem.class);
        when(item.getItemId()).thenReturn(0);

        boolean result = presenter.onOptionsItemSelected(item);

        assertThat(result).isEqualTo(false);
        verifyNoMoreInteractions(viewable);
    }

    @Test
    public void testOnMockModeCheckedChanged() {
        // TODO create tests for PrefUtils
        PowerMockito.mockStatic(PrefUtils.class);

        presenter.onMockModeCheckedChanged(true);

        verifyNoMoreInteractions(viewable);
    }
}

File: app/src/test/java/com/leaguechampions/ui/settings/SettingsPresenterTest.kt
package com.leaguechampions.ui.settings

import com.leaguechampions.BuildConfig
import com.leaguechampions.utils.PrefUtils
import org.junit.After
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.powermock.core.classloader.annotations.PrepareForTest
import org.powermock.modules.junit4.PowerMockRunner

@RunWith(PowerMockRunner::class)
@PrepareForTest(BuildConfig::class, PrefUtils::class)
class SettingsPresenterTest {

//    private val fieldBuildType = "buildType"
//
//    private lateinit var presenter: SettingsPresenter
//
//    @Mock private lateinit var view: SettingsPresenter.SettingsView
//    @Mock private lateinit var sharedPreferences: SharedPreferences
//
    @Before
    fun setUp() {
//        presenter = SettingsPresenter(view, sharedPreferences)
    }

    @After
    fun tearDown() {

    }

    @Test
    fun delete_me() {

    }

//    @Test
//    fun testOnActivityCreated() {
//        val mockMode = true
//        PowerMockito.mockStatic(PrefUtils::class.java)
//        `when`(PrefUtils.isMockMode(sharedPreferences)).thenReturn(mockMode)
//        ReflectionUtils.setField(presenter, fieldBuildType, "debug")
//
//        presenter.onActivityCreated(null, null)
//
//        verify(view).setVersion(BuildConfig.VERSION_NAME)
//        verify(view).showDeveloperOptions()
//        verify(view).showMockMode(mockMode)
//        verifyNoMoreInteractions(view)
//    }
//
//    @Test
//    fun testOnActivityCreated_WhenBuildTypeRelease_OnlySetsVersion() {
//        PowerMockito.mockStatic(PrefUtils::class.java)
//        ReflectionUtils.setField(presenter, fieldBuildType, "release")
//
//        presenter.onActivityCreated(null, null)
//
//        verify(view).setVersion(BuildConfig.VERSION_NAME)
//        verifyNoMoreInteractions(view)
//    }
//
//    @Test
//    fun testOnOptionsItemSelected() {
//        val item = mock(MenuItem::class.java)
//        `when`(item.itemId).thenReturn(android.R.id.home)
//
//        val result = presenter.onOptionsItemSelected(item)
//
//        assertThat(result).isEqualTo(true)
//        verify(view).doFinish()
//        verifyNoMoreInteractions(view)
//    }
//
//    @Test
//    fun testOnOptionsItemSelected_WhenItemNotHome_DoNothing() {
//        val item = mock(MenuItem::class.java)
//        `when`(item.itemId).thenReturn(0)
//
//        val result = presenter.onOptionsItemSelected(item)
//
//        assertThat(result).isEqualTo(false)
//        verifyNoMoreInteractions(view)
//    }
//
//    @Test
//    fun testOnMockModeCheckedChanged() {
//        val mockMode = true
//        val editor = mock(SharedPreferences.Editor::class.java)
//        PowerMockito.mockStatic(PrefUtils::class.java)
//        `when`(sharedPreferences.edit()).thenReturn(editor)
//        `when`(editor.putBoolean(Const.PREF_KEY_MOCK_MODE, mockMode)).thenReturn(editor)
//
//        presenter.onMockModeCheckedChanged(mockMode)
//
//        verifyNoMoreInteractions(view)
//    }
}
